### PR TITLE
Add design diagrams

### PR DESCRIPTION
**Pull request by Firstyear**
Relates #1481 - start to diagram out possible ways to implement access controls

Checklist

- [ ] This pr contains no AI generated code
- [ ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
You lilkely need to view this as raw.

**Comment by yaleman** [comment]
I might have turned it into mermaid because ... 😄

File: book/src/developers/designs/domain_join.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+In this case when a user authenticates with ssh, since we are a trusted machine we should be able to
+request short-term and limited credentials on the users behalf.
+
+An example is that we could dynamicly request TLS certificates or kerberos credentials.
+
+Normally with ssh in this manner, everything has to use kerberos. This would force users to kinit on
+their machine to ssh and forward their credentials to the next machine. This causes friction since
+configuring kerberos on machines is an exercise in frustration, and with BYOD it gets even worse. In
+addition if a use ssh's with an ssh key, the only viable kinit mechanism is password or password + totp. Both
```
Comment (dmulder) [inline book/src/developers/designs/domain_join.md:41]: You've got a typo here.

File: book/src/developers/designs/domain_join.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+Machines may re-enroll with an alternate token.
+
+The join token should be signed. The JWK pub key should be available at a known HTTPS uri so that the client
+can use it to validate the join token and it's content. This *may* allow policy to be embedded into
+the join token for the client to self-adhere to in the join process.
+
+### Machine Auditing
+
+The machine should be audited to be in a secure state. It's not yet clear how to proceed here, but
+we should consider using TPM PCRs with secure boot to measure this and validate the machine state.
```
Comment (dmulder) [inline book/src/developers/designs/domain_join.md:86]: Azure Intune clients for example would check here if the disk is encrypted.

File: book/src/developers/designs/domain_join.md (hunk 1 above)
Comment (yaleman) [reply book/src/developers/designs/domain_join.md:41]: Could you be slightly more specific than "a typo here" while highlighting an entire paragraph? 😄

File: book/src/developers/designs/domain_join.md (hunk 2 above)
Comment (Firstyear) [reply book/src/developers/designs/domain_join.md:86]: Yeah. We need to see how that's working in ALP and other distros generally. Not every distro does FDE with TPM's, meaning that FDE != passing secure boot on all platforms. ALP however is TPM PCR secure boot based, so we can use that as a measure of "some" boot health. I'm also not sure how to programmatically check that either.

File: book/src/developers/designs/domain_join.md (hunk 1 above)
Comment (Firstyear) [reply book/src/developers/designs/domain_join.md:41]: I spotted the typo.

File: book/src/developers/designs/domain_join.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+The machine must submit it's public key, the time value and the signature. This should accompany the
+join token.
+
+If the signature is valid, and the join token is correct, then the machine is joined and has a
+machine account created. The machine account is linked to the join token.
+
+### Machine Account
+
+The machine account is a new form of account, similar to a service account. It should identify the
+machine, it's hostname, and other properties. It should also contain the machines public key id.
```
Comment (yaleman) [inline book/src/developers/designs/domain_join.md:124]:
```suggestion
machine, its hostname, and other properties. It should also contain the machine's public key id.
```

File: book/src/developers/designs/domain_join.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+### Public Key Submission
+
+The private key should be generated and stored in a TPM/HSM. If possible, we should also submit
+attestation of this.
+
+The submission of the public key should prevent replays, and should sign either a nonce or the
+current time. The current time must be valid to within a number of seconds. The nonce must be
+created by the server.
+
+The machine must submit it's public key, the time value and the signature. This should accompany the
```
Comment (yaleman) [inline book/src/developers/designs/domain_join.md:115]:
```suggestion
The machine must submit its public key, the time value and the signature. This should accompany the
```

File: book/src/developers/designs/domain_join.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+Join tokens are persisted in the database allowing tracing back to the usage of the token.
+
+Every machine that is joined by that token will related back to that token. This allows auditing of
+which token was used to join which machine.
+
+Machines may re-enroll with an alternate token.
+
+The join token should be signed. The JWK pub key should be available at a known HTTPS uri so that
+the client can use it to validate the join token and it's content. This _may_ allow policy to be
```
Comment (yaleman) [inline book/src/developers/designs/domain_join.md:93]:
```suggestion
the client can use it to validate the join token and its content. This _may_ allow policy to be
```

File: book/src/developers/designs/domain_join.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+## Overview
+
+Since the machine would now be an entity requiring authentication, we need to have a process to
+establish and maintain this trust relationship.
+
+1. A join token is created by a user who is authorised to perform domain joins.
+2. The machine is audited for a known trust state. This process may vary from site to site. A future
+   improvement could be that the join token can only release on certain TPM PCR values.
+3. The join token is yielded to the kanidm unix daemon which submits it's signing key to the kanidm
```
Comment (yaleman) [inline book/src/developers/designs/domain_join.md:68]:
```suggestion
3. The join token is yielded to the Kanidm UNIX daemon which submits its signing key to the Kanidm
```

File: book/src/developers/designs/domain_join.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+When a user logs in to a machine, it may be required that they can use that authentication to
+identify themself to other systems. When a user authenticates with credentials such as ssh-keys,
+it's not possible to use these to request other forwardable credentials - and ssh agent forwarding
+only allows forwarding of ssh credentials, not other types of credentials that may be needed.
+
+In this case when a user authenticates with ssh, since we are a trusted machine we should be able to
+request short-term and limited credentials on the users behalf.
+
+An example is that we could dynamicly request TLS certificates or kerberos credentials.
```
Comment (yaleman) [inline book/src/developers/designs/domain_join.md:42]:
```suggestion
An example is that we could dynamically request TLS certificates or Kerberos credentials.
```

**Review by yaleman — CHANGES_REQUESTED** [review]
My usual yak-shaving stuff, but I can't find the new onboarding role that was recently added in the diagrams?

**Comment by Firstyear** [comment]
> My usual yak-shaving stuff, but I can't find the new onboarding role that was recently added in the diagrams?

I may not have added it here yet. Let me check that.

**Comment by Firstyear** [comment]
@yaleman There you go :)

**Review by yaleman — APPROVED** [review]

**Comment by yaleman** [comment]
Seems legit so far, I'm looking forward to the endpoint data so I can finally use it to power my ansible inventory ^_^

**Pull request merged**